teuthology/orchestra/run.py

```diff
@@ -221,6 +233,46 @@ def __repr__(self):
             name=self.hostname,
             )
 
+class ErrorScanner(object):
+    """
+    Scan for unit tests errors in teuthology.log 
+    """
+    __flag__ = 0
+    def __init__(self, test_names=[]):
+        self.test_names = test_names
+
+    def scan(self):
+        logfile = self.__logfile__
+        if not logfile or not self.test_names:
+            return None
+        
+        ERROR_PATTERN = {
+            "nose": r"ERROR:\s",
+            "gtest":  r"\[\s\sFAILED\s\s\]",
+        }
+        logs = []
+        with open(logfile, 'r') as f:
+            logs = f.readlines()
```

Comment:
Rather than read the entire file at once - which could use large amounts of memory - let's look into using a [generator](https://wiki.python.org/moin/Generators) to read and process one line (or several lines) at a time. That could be integrated with the `for` loop below.

Reply:
Great point! For lazy reading of files, we can simply loop through the file object created with `open('textfile')` like:
```python
for line in open("file.txt"):
     print(line)
```
This will not load the entire file in memory, instead read it line by line. 

For `__flag__` implementation, we can use [seek() & tell()](https://docs.python.org/3/tutorial/inputoutput.html#methods-of-file-objects) methods, to do something like: 
(reading files top to bottom here)
```python
__flag__ = 3   # last read byte

f = open('sample.txt', 'r')
f.seek(__flag__ + 1) # seek to next unread byte

for line in f:
    # search regex matches in `line`

__flag__ = f.tell()

f.close()
```
Thoughts about this approach?

Reply:
After exploring the following ways to read logs, I compared them with [timeit](https://docs.python.org/3/library/timeit.html) and [tracemalloc](https://docs.python.org/3/library/tracemalloc.html#tracemalloc.get_traced_memory) on a 0.5GB file.
1. `f.readline()` 
2.  [islice()](https://stackoverflow.com/a/6335876)
3.  f.readlines(num_of_lines)
4.  `for line in f`
5.  [mmap](https://docs.python.org/3/library/mmap.html)

Looks like for-in loop on a file object (option 4 above) is the most memory and time efficient for huge log files. 
Further, re.compile improved the performance a lot.